// @bit/adeoy.utils.data-table/dist/Columns.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _sortAsc = _interopRequireDefault(require("./assets/img/sort-asc.svg"));

var _sortDesc = _interopRequireDefault(require("./assets/img/sort-desc.svg"));

var _unsorted = _interopRequireDefault(require("./assets/img/unsorted.svg"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var Columns = function Columns(_ref) {
  var columns = _ref.columns,
      currentSortedCol = _ref.currentSortedCol,
      handleSortColumn = _ref.handleSortColumn;
  return /*#__PURE__*/_react.default.createElement("thead", null, /*#__PURE__*/_react.default.createElement("tr", null, columns.map(function (col, i) {
    var nextSort = 1;
    var iconSorted = _unsorted.default;

    if (i === currentSortedCol.idx) {
      if (currentSortedCol.sortedBy === 1) {
        iconSorted = _sortAsc.default;
        nextSort = -1;
      } else if (currentSortedCol.sortedBy === -1) {
        iconSorted = _sortDesc.default;
      }
    }

    return /*#__PURE__*/_react.default.createElement("th", {
      key: i,
      onClick: function onClick() {
        return handleSortColumn(i, nextSort);
      },
      style: {
        cursor: 'pointer'
      }
    }, col.title, " ", /*#__PURE__*/_react.default.createElement("img", {
      src: iconSorted,
      alt: "Sort ".concat(col.title),
      width: "16px"
    }));
  })));
};

var _default = Columns;
exports.default = _default;
