// @bit/adeoy.utils.data-table/dist/index.js
"use strict";

function _typeof(obj) { "@babel/helpers - typeof"; if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireWildcard(require("react"));

var _reactstrap = require("reactstrap");

var _Columns = _interopRequireDefault(require("./Columns"));

var _Rows = _interopRequireDefault(require("./Rows"));

var _Pagination = _interopRequireDefault(require("./Pagination"));

var _Filter = _interopRequireDefault(require("./Filter"));

var _PageSize = _interopRequireDefault(require("./PageSize"));

var _Counting = _interopRequireDefault(require("./Counting"));

var _utils = require("./utils");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function _getRequireWildcardCache() { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _toConsumableArray(arr) { return _arrayWithoutHoles(arr) || _iterableToArray(arr) || _unsupportedIterableToArray(arr) || _nonIterableSpread(); }

function _nonIterableSpread() { throw new TypeError("Invalid attempt to spread non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); }

function _iterableToArray(iter) { if (typeof Symbol !== "undefined" && Symbol.iterator in Object(iter)) return Array.from(iter); }

function _arrayWithoutHoles(arr) { if (Array.isArray(arr)) return _arrayLikeToArray(arr); }

function _slicedToArray(arr, i) { return _arrayWithHoles(arr) || _iterableToArrayLimit(arr, i) || _unsupportedIterableToArray(arr, i) || _nonIterableRest(); }

function _nonIterableRest() { throw new TypeError("Invalid attempt to destructure non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); }

function _unsupportedIterableToArray(o, minLen) { if (!o) return; if (typeof o === "string") return _arrayLikeToArray(o, minLen); var n = Object.prototype.toString.call(o).slice(8, -1); if (n === "Object" && o.constructor) n = o.constructor.name; if (n === "Map" || n === "Set") return Array.from(o); if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen); }

function _arrayLikeToArray(arr, len) { if (len == null || len > arr.length) len = arr.length; for (var i = 0, arr2 = new Array(len); i < len; i++) { arr2[i] = arr[i]; } return arr2; }

function _iterableToArrayLimit(arr, i) { if (typeof Symbol === "undefined" || !(Symbol.iterator in Object(arr))) return; var _arr = []; var _n = true; var _d = false; var _e = undefined; try { for (var _i = arr[Symbol.iterator](), _s; !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i["return"] != null) _i["return"](); } finally { if (_d) throw _e; } } return _arr; }

function _arrayWithHoles(arr) { if (Array.isArray(arr)) return arr; }

var DataTable = function DataTable(props) {
  var data = props.data || [];
  var columns = props.columns;
  var onClickRow = props.onClickRow || null;

  var _useState = (0, _react.useState)(data),
      _useState2 = _slicedToArray(_useState, 2),
      dataFiltered = _useState2[0],
      setDataFiltered = _useState2[1];

  var _useState3 = (0, _react.useState)(props.pageSize || 10),
      _useState4 = _slicedToArray(_useState3, 2),
      pageSize = _useState4[0],
      setPageSize = _useState4[1];

  var _useState5 = (0, _react.useState)(Math.ceil(data.length / (props.pageSize || 10))),
      _useState6 = _slicedToArray(_useState5, 2),
      maxPage = _useState6[0],
      setMaxPage = _useState6[1];

  var _useState7 = (0, _react.useState)(1),
      _useState8 = _slicedToArray(_useState7, 2),
      currentPage = _useState8[0],
      setCurrentPage = _useState8[1];

  var _useState9 = (0, _react.useState)(data.slice(0, 10)),
      _useState10 = _slicedToArray(_useState9, 2),
      currentRows = _useState10[0],
      setCurrentRows = _useState10[1];

  var _useState11 = (0, _react.useState)(""),
      _useState12 = _slicedToArray(_useState11, 2),
      filter = _useState12[0],
      setFilter = _useState12[1];

  var _useState13 = (0, _react.useState)({
    idx: -1,
    sortedBy: 0
  }),
      _useState14 = _slicedToArray(_useState13, 2),
      currentSortedCol = _useState14[0],
      setCurrentSortedCol = _useState14[1];

  var handlePageSize = function handlePageSize(val) {
    setPageSize(val);
    setMaxPage(Math.ceil(dataFiltered.length / val));
    setCurrentPage(1);
    handleDataFiltered(1, val, dataFiltered);
  };

  var handleFilter = function handleFilter(val) {
    setFilter(val);
    var query = "%" + val.toLowerCase().replace(new RegExp(" ", "g"), "%") + "%";
    var results = [];

    var _loop = function _loop(i) {
      var row = data[i];

      if (val === "") {
        results.push(row);
        return "continue";
      }

      var searchRow = columns.map(function (col) {
        if (col.data) {
          return (0, _utils.resolveObjProp)(col.data, row);
        } else {
          return '';
        }
      });
      var cols = Object.values(searchRow).join(" ");

      if ((0, _utils.like)(cols, query)) {
        results.push(row);
      }
    };

    for (var i = 0; i < data.length; i++) {
      var _ret = _loop(i);

      if (_ret === "continue") continue;
    }

    setDataFiltered(results);
    setMaxPage(Math.ceil(results.length / pageSize));
    setCurrentPage(1);
    handleDataFiltered(1, pageSize, results);
  };

  var handleCurrentPage = function handleCurrentPage(page) {
    setCurrentPage(page);
    handleDataFiltered(page, pageSize, dataFiltered);
  };

  var handleDataFiltered = function handleDataFiltered(currentPage, pageSize, dataFiltered) {
    var a = (currentPage - 1) * pageSize;
    var z = currentPage * pageSize;
    setCurrentRows(dataFiltered.slice(a, z));
  };

  var handleSortColumn = function handleSortColumn(i, nextSort) {
    var data = _toConsumableArray(dataFiltered.sort(function (a, b) {
      var col = columns[i];
      var keyA = '',
          keyB = '';

      if (col.data) {
        keyA = (0, _utils.resolveObjProp)(col.data, a);
        keyB = (0, _utils.resolveObjProp)(col.data, b);
      } else {
        keyA = col.format(a);
        keyB = col.format(b);
      }

      if (keyA < keyB) return -nextSort;
      if (keyA > keyB) return nextSort;
      return 0;
    }));

    setDataFiltered(data);
    setCurrentRows(data.slice(0, 10));
    setCurrentPage(1);
    setCurrentSortedCol({
      idx: i,
      sortedBy: nextSort
    });
  };

  (0, _react.useEffect)(function () {
    setDataFiltered(data);
    setCurrentRows(data.slice(0, 10));
  }, [data]);
  return /*#__PURE__*/_react.default.createElement(_react.Fragment, null, /*#__PURE__*/_react.default.createElement(_reactstrap.Row, {
    className: "mb-2"
  }, /*#__PURE__*/_react.default.createElement(_reactstrap.Col, {
    xs: "6",
    md: "4"
  }, /*#__PURE__*/_react.default.createElement(_PageSize.default, {
    pageSize: pageSize,
    handlePageSize: handlePageSize
  })), /*#__PURE__*/_react.default.createElement(_reactstrap.Col, {
    xs: "6",
    md: {
      size: 4,
      offset: 4
    }
  }, /*#__PURE__*/_react.default.createElement(_Filter.default, {
    filter: filter,
    handleFilter: handleFilter
  }))), /*#__PURE__*/_react.default.createElement(_reactstrap.Row, null, /*#__PURE__*/_react.default.createElement(_reactstrap.Col, {
    xs: "12"
  }, /*#__PURE__*/_react.default.createElement(_reactstrap.Table, {
    striped: props.striped,
    hover: props.hover,
    bordered: props.bordered,
    borderless: props.borderless,
    responsive: props.responsive
  }, /*#__PURE__*/_react.default.createElement(_Columns.default, {
    columns: columns,
    currentSortedCol: currentSortedCol,
    handleSortColumn: handleSortColumn
  }), /*#__PURE__*/_react.default.createElement(_Rows.default, {
    columns: columns,
    currentRows: currentRows,
    onClickRow: onClickRow
  })))), /*#__PURE__*/_react.default.createElement(_reactstrap.Row, null, /*#__PURE__*/_react.default.createElement(_reactstrap.Col, {
    xs: "12",
    md: "6"
  }, /*#__PURE__*/_react.default.createElement(_Counting.default, {
    currentPage: currentPage,
    pageSize: pageSize,
    currentMax: dataFiltered.length
  })), /*#__PURE__*/_react.default.createElement(_reactstrap.Col, {
    xs: "12",
    md: "6",
    className: "text-right"
  }, /*#__PURE__*/_react.default.createElement(_Pagination.default, {
    currentPage: currentPage,
    handleCurrentPage: handleCurrentPage,
    maxPage: maxPage
  }))));
};

var _default = DataTable;
exports.default = _default;
